File: 875.cpp
#include <algorithm>
#include <cmath>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
  int minEatingSpeed(vector<int> &piles, int h) {
    int l = 1;
    int r = *max_element(piles.begin(), piles.end());
    int lastWorkingElt = -1;
    while (l <= r) {
      int currentK = (l + r) / 2;
      int currentHours = 0;
      for (int pile : piles) {
        currentHours += ceil((double)pile / (double)currentK);
      }
      if (currentHours <= h) {
        r = currentK - 1;
        lastWorkingElt = currentK;
      } else {
        l = currentK + 1;
      }
    }
    return lastWorkingElt;
  }
};

int main() {
  vector<int> t1 = {30, 11, 23, 4, 20};
  Solution s;
  int res = s.minEatingSpeed(t1, 5);
  cout << res << endl;
  cout << "expected: " << 23 << endl;
}

File: 875.py
from math import ceil


class Solution:
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        # return true if within h hours, at a rate of k bananas per hour,
        # all bananas can be consumed. False otherwise.

        def helper(piles, k, h):
            total = 0
            for pile in piles:
                total += ceil(pile / k)
                if total > h:
                    return False
            return True

        low = 1
        high = max(piles)
        lastWorkingK = -1
        while low <= high:
            currentK = (high - low) // 2 + low
            if helper(piles, currentK, h):
                high = currentK - 1
                lastWorkingK = currentK
            else:
                low = currentK + 1
        return lastWorkingK


s = Solution()
print(s.minEatingSpeed([30, 11, 23, 4, 20], 6))
print("expected 23")
